app/gui2/src/components/GraphEdge.vue

```diff
@@ -1,57 +1,501 @@
+      const j0BeyondSource = Math.abs(inputs.targetOffset.x) + RADIUS_MAX * 2
+      const j0BeyondTarget = Math.abs(sourceX) + RADIUS_MAX
+      j0x = Math.abs(Math.max(j0BeyondTarget, j0BeyondSource)) * signX
+      heightAdjustment = 0
+    }
+    if (j0x == null || j1x == null || heightAdjustment == null) return null
+    const attachmentHeight = targetMaxAttachmentHeight ?? 0
+    const top = Math.min(
+      inputs.targetOffset.y - MIN_APPROACH_HEIGHT - attachmentHeight + heightAdjustment,
+      0,
+    )
+    const source = new Vec2(sourceX, 0)
+    const j0 = new Vec2(j0x, top / 2)
+    const j1 = new Vec2(j1x, top)
+    // The corners meet the target attachment at the top of the node.
+    const attachmentTarget = attachment?.target ?? inputs.targetOffset
+    return {
+      points: [source, j0, j1, attachmentTarget],
+      maxRadius: RADIUS_MAX,
+      targetAttachment: attachment,
+    }
+  }
+}
+
+type Line = { axis: 'h' | 'v'; length: number }
+type Arc = { radius: number; signX: number; signY: number; sweep: 0 | 1 }
+type Element = Arc | Line
+
+function pathElements(junctions: JunctionPoints): { start: Vec2; elements: Element[] } {
+  const elements: Element[] = []
+  const pushLine = (line: Line) => {
+    if (line.length === 0) return
+    const e = elements.pop()
+    if (e != null) {
+      if ('axis' in e && e.axis == line.axis) {
+        e.length += line.length
+        elements.push(e)
+      } else {
+        elements.push(e)
+        elements.push(line)
+      }
+    } else {
+      elements.push(line)
+    }
+  }
+  const start = junctions.points[0]
+  if (start == null) return { start: Vec2.Zero(), elements: [] }
+  let prev = start
+  junctions.points.slice(1).map((j, i) => {
+    const d = j.sub(prev)
+    const radius = Math.min(junctions.maxRadius, Math.abs(d.x), Math.abs(d.y))
+    const signX = Math.sign(d.x)
+    const signY = Math.sign(d.y)
+    const dx = (Math.abs(d.x) - radius) * signX
+    const dy = (Math.abs(d.y) - radius) * signY
+    const h: Line = { axis: 'h', length: dx }
+    const v: Line = { axis: 'v', length: dy }
+    const sweep = (signX === signY) === (i % 2 === 0) ? 1 : 0
+    if (i % 2 == 0) {
+      pushLine(h)
+      elements.push({ radius, signX, signY, sweep })
+      pushLine(v)
+    } else {
+      pushLine(v)
+      elements.push({ radius, signX, signY, sweep })
+      pushLine(h)
+    }
+    prev = j
+  })
+  return { start, elements }
+}
+
+function render(sourcePos: Vec2, elements: Element[]): string {
+  let out = `M ${sourcePos.x} ${sourcePos.y}`
+  for (const e of elements) {
+    if ('axis' in e) {
+      out += ` ${e.axis} ${e.length}`
+    } else {
+      const dx = e.radius * e.signX
+      const dy = e.radius * e.signY
+      out += ` a ${e.radius} ${e.radius} 0 0 ${e.sweep} ${dx} ${dy}`
+    }
+  }
+  return out
+}
+
+const currentJunctionPoints = computed(() => {
+  const target_ = targetPos()
+  const source_ = sourcePos()
```

Comment:
The target and source pos should be computed values, since those are used multiple times.